### PR TITLE
Refactor: `CrawlerBase` and `PermissionsManager` snapshotting pattern

### PR DESCRIPTION
**Pull request by asnare**
## Changes

This PR:

 - Refactors the existing permissions crawler so that it uses the same fetch/crawl/snapshot pattern as the rest of the crawlers.
 - Removes the permissions crawler's bespoke `.cleanup()` implementation in favour of the `.reset()` method already available on the base class.
 - Some dead code was removed on the permissions crawler: `.load_for_all()`
 - Specifies `.snapshot()` as an interface (on `CrawlerBase`) that all crawlers need to support.

### Linked issues

Progresses #2074, by laying the ground for crawlers to support refreshing.

Note that this PR and #2392 have a behaviour conflict that will trigger a failure in `test_runtime_crawl_permissions` if either is merged: resolving this is trivial.

### Functionality

- modified existing workflow: `assessment`

### Tests

- [X] updated unit tests
- [X] updated integration tests

### CODE REVIEW

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 1]
```diff
@@ -431,11 +432,9 @@ def __init__(  # pylint: disable=too-many-arguments
         self._external_id_match = external_id_match
         self._verify_timeout = verify_timeout
 
-    def snapshot(self) -> list[MigratedGroup]:
-        return self._snapshot(self._fetcher, self._crawler)
-
     def has_groups(self) -> bool:
-        return len(self.snapshot()) > 0
+        first_element = islice(self.snapshot(), 1)
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:436]: can you revert the change on this method? i find it risky. it's fine to load snapshot twice, as it's materialized

**Review by nfx — CHANGES_REQUESTED** [review]
one tiny change

File: src/databricks/labs/ucx/workspace_access/groups.py (hunk 1 above)
Comment (asnare) [reply src/databricks/labs/ucx/workspace_access/groups.py]: I've reverted it.

This was intended to be a poor-man's `limit 1` because `.snapshot()` is normally lazy, etc.

Comment (asnare) [reply src/databricks/labs/ucx/workspace_access/groups.py]: Pfft. And I just noticed the method isn't actually called from anywhere (including tests). So I'm just going to nuke it entirely.

Comment (asnare) [reply src/databricks/labs/ucx/workspace_access/groups.py]: … and removed.

**Review by nfx — APPROVED** [review]
lgtm
